Replace `Identifiers* names` with `ArgumentLabels* names` across the frontend

Ay1nDas opened the pull request:
This PR is a continuation of PR #21262. Check it out for more details.

### #Summary
This PR replaces all uses of the `Identifiers* names` field in expression-related functions (`callExp`, `newExp`, etc.) with a more expressive and structured alternative: `ArgumentLabels* argLabels`.

- Replaced all `names` fields (of type `Identifiers*`) with `argLabels` (of type `ArgumentLabels* = Array!(ArgumentLabel)`).
- Code using `argLabels` is backwards-compatible where necessary by providing `Identifiers*` via helper conversion.

ibuclaw commented on compiler/src/dmd/cxxfrontend.d:
```diff
@@ -256,9 +256,9 @@ Expression getDefaultValue(EnumDeclaration ed, Loc loc)
  */
-void expandTuples(Expressions* exps, Identifiers* names = null)
+void expandTuples(Expressions* exps, ArgumentLabels* argLabels = null)
 {
-    return dmd.expression.expandTuples(exps, names);
+    return dmd.expression.expandTuples(exps, argLabels);
 }
```
I don't understand how is this passing the testsuite. Are new symbols being added without link tests?

ibuclaw commented:
This shouldn't be targeting stable unless it's a bug fix. Thanks.

Ay1nDas commented:
Apologies for the confusion caused. I forgot to mention this is a continuation of PR #21262, in that PR I added a new `struct ArgLabel` in [`dmd/expression.d`](https://github.com/dlang/dmd/pull/21262/files#diff-36992bf6781096796c127b16ffe56bee2f8950049796e53f5b856223a1fd2a9eR3291-R3295) and implemented it in some places to fix issue #21215. 
This PR is to replace `ArgumentLabels` ([defined here](https://github.com/dlang/dmd/pull/21262/files#diff-08e673c291c756dd7ea412c2fed539e914772ef4921b2601de5d3400d25544f4R32)) so that we don't have to maintain 3 different arrays for expressions (arguments, argLabels, names).
#### Now to answer the questions:
- I modified all the required instances of `expandTuples()` to accept the new argLabels. You can see the function definition at [dmd/expression.d](https://github.com/dlang/dmd/pull/21297/files#diff-36992bf6781096796c127b16ffe56bee2f8950049796e53f5b856223a1fd2a9eR90-R166). I think that's why it is passing the tests.
- `ArgumentLabel` is defined [here](https://github.com/dlang/dmd/pull/21262/files#diff-36992bf6781096796c127b16ffe56bee2f8950049796e53f5b856223a1fd2a9eR3291-R3296). `ArgLabels` is an `alias` defined [here](https://github.com/dlang/dmd/pull/21262/files#diff-08e673c291c756dd7ea412c2fed539e914772ef4921b2601de5d3400d25544f4R32).
- This pr is targeting `stable` instead of `master` as the definition for `ArgLabel` was added in `stable` after the merge of PR #21262. Which itself was a bug fix for issue #21215. 
##### I hope this clarifies everything. I split the changes into 2 PRs as I thought it would be confusing that... why all the `names` are being replaced with `argLabels` to fix a bug?

dkorpel commented on compiler/src/dmd/expression.d:
```diff
@@ -3265,41 +3267,55 @@ extern (C++) final class DotTypeExp : UnaExp
 
+//Gets all the names of the argument labels from a list of expressions
+// and returns them in a new list. The list is not owned by the caller.
+ Identifiers* extractNames(ArgumentLabels* argLabels)
```
So this function exists to pass argLabels to `resolveStructLiteralNamedArgs` which takes a separate Identifier[] for names and delegate to get locations. This is al getting really clumsy, can you make `resolveStructLiteralNamedArgs` just take an `ArgumentLabels` parameter?

Ay1nDas replied:
I tried to replace names with argLabels in `resolveStructLiteralNamedArgs`. But the following causes the problem:
https://github.com/dlang/dmd/blob/e715eb845977a85ecfc27d257b31ecf16fea0719/compiler/src/dmd/initsem.d#L174

here the `i.field[]` is of type `identifier[]`. And, `.field` is defined here:
https://github.com/dlang/dmd/blob/e715eb845977a85ecfc27d257b31ecf16fea0719/compiler/src/dmd/astbase.d#L6633-L6636

The StructInitializer class was added 8 years ago with PR #6836. And it looks deeply integrated.
I tried to replace the `Identifiers field` with `ArgumentLabels argField`, but I couldn't solve the errors that it creates.
So, I took the easy route. 
Is there any workaround that I can do? Like adding a false loc `Loc.initial` to the field before passing?

dkorpel replied:
You could pass a delegate `(size_t j) => i.field[j]` and `(size_t j) => i.names[j]` similar to loc, or use a template parameters.

dkorpel reviewed (changes requested):
This PR's diff does 2 things:
- rename `names` to `argLabels`
- fix the bug where argLabels don't get expanded with tuples

Mixing refactoring and bug fixes in a single pull requests should generally be avoided, because when a regression happens and a bissection points to this PR, it is still hard to see what's going on because actual code changes are mixed with many lines that simply rename something. Can you simplify this PR by applying the code changes to the original `names` variable?

Ay1nDas committed:
replacing identifiers* with ArgumentLabels*

dkorpel commented on compiler/src/dmd/expression.d:
```diff
@@ -81,42 +81,43 @@ inout(Expression) lastComma(inout Expression e)
+ *     argLabels = optional array of argumentLabels(name and it's location) corresponding to Expressions
  */
-void expandTuples(Expressions* exps, Identifiers* names = null)
+void expandTuples(Expressions* exps, ArgumentLabels* argLabels = null)
```
Something to that effect. Rename `argLabels` to `names` in this PR, and rename `names` to `argLabels` in a follow-up PR.

Ay1nDas replied:
Changing `expandTuples` back fails many tests in `tests/fail_compilation`. So, I have 2 solutions for it.
- Merging the PR with all failed tests. And as soon as it gets merged, creating a new issue and implementing the `expandTupples()` changes.
- Applying the `expandTuples ()` changes to fix the bug before this PR. But it adds both `argLabels` and `names` to the  function which creates redundant Code.

Please let me which is more suitable.

dkorpel replied:
I'm suggesting taking the PR as it is now, and renaming `argLabels` to `names`

Ay1nDas replied:
Apologies, are you saying that I make it `ArgumentLabels* names`?

dkorpel replied:
`ArgumentLabels* names`

dkorpel replied:
By the way, if it's too much effort / doesn't help much, we can just merge this PR anyway.

Ay1nDas replied:
I have a question... Do you mean replace all the `argLabels` with `names` across the frontend or just in the `expandTuples()`.

dkorpel replied:
Everywhere, so all the diff lines that are simply renames disappear

dkorpel commented on compiler/src/dmd/expression.d:
```diff
@@ -3271,36 +3272,37 @@ extern (C++) final class DotTypeExp : UnaExp
     }
 }
 
+// A list of argument labels
```
It's not a list though ;)

Ay1nDas replied:
Should I change it to an array of arguments?

dkorpel replied:
I just mean the comment is wrong, it describes `struct ArgumentLabel` as a list of argument labels, while the structure represents a single argument label.

dkorpel replied:
It now says "An array of argument labels" which is still describing a single structure as multiple of them

Ay1nDas commented on compiler/src/dmd/expression.d:
```diff
@@ -3271,36 +3272,36 @@ extern (C++) final class DotTypeExp : UnaExp
     }
 }
 
+// Contains both `name` and `location of the name` for an expression.
```
Do you think it's better now?

dkorpel replied:
Yes, thanks!
